plugins/modem-manager/fu-mm-device.c

```diff
@@ -0,0 +1,668 @@
+					       file_info->bytes,
+					       &iterate_context.error)) {
+			g_warning ("Failed to write file '%s': %s",
+				   file_info->filename, iterate_context.error->message);
+			break;
+		}
+	}
+
+	if (iterate_context.error != NULL) {
+		g_propagate_error (error, iterate_context.error);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_mm_device_write_firmware (FuDevice *device, GBytes *fw, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE (device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(FuArchive) archive = NULL;
+	g_autoptr(GPtrArray) array = NULL;
+
+	/* lock device */
+	locker = fu_device_locker_new (device, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* qmi pdc write operation */
+	if (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_QMI_PDC)
+		return fu_mm_device_write_firmware_qmi_pdc (device, fw, error);
+
+	g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED,
+		     "unsupported update method");
+	return FALSE;
+}
+
+static gboolean
+fu_mm_device_attach (FuDevice *device, GError **error)
+{
+	g_autoptr(FuDeviceLocker) locker  = NULL;
+
+	/* lock device */
+	locker = fu_device_locker_new (device, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* wait for re-probing after uninhibiting */
+	fu_device_set_remove_delay (device, FU_MM_DEVICE_REMOVE_DELAY_REPROBE);
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static void
+fu_mm_device_init (FuMmDevice *self)
+{
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_USE_RUNTIME_VERSION);
+	fu_device_set_summary (FU_DEVICE (self), "Mobile broadband device");
+	fu_device_add_icon (FU_DEVICE (self), "network-modem");
+}
+
+static void
+fu_mm_device_finalize (GObject *object)
+{
+	FuMmDevice *self = FU_MM_DEVICE (object);
+	g_object_unref (self->manager);
+	if (self->omodem != NULL)
+		g_object_unref (self->omodem);
+	g_free (self->detach_fastboot_at);
+	g_free (self->port_at);
+	g_free (self->port_qmi);
+	g_free (self->inhibition_uid);
+	G_OBJECT_CLASS (fu_mm_device_parent_class)->finalize (object);
+}
+
+static void
+fu_mm_device_class_init (FuMmDeviceClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (klass);
+	FuDeviceClass *klass_device = FU_DEVICE_CLASS (klass);
+	object_class->finalize = fu_mm_device_finalize;
+	klass_device->to_string = fu_mm_device_to_string;
+	klass_device->probe = fu_mm_device_probe;
+	klass_device->detach = fu_mm_device_detach;
+	klass_device->write_firmware = fu_mm_device_write_firmware;
+	klass_device->attach = fu_mm_device_attach;
+}
+
+FuMmDevice *
+fu_mm_device_new (MMManager *manager, MMObject *omodem)
+{
+	FuMmDevice *self = g_object_new (FU_TYPE_MM_DEVICE, NULL);
+	self->manager = g_object_ref (manager);
+	self->omodem = g_object_ref (omodem);
+	self->port_at_ifnum = -1;
+	return self;
+}
+
+FuMmDevice *
+fu_mm_device_udev_new (MMManager	*manager,
```

Comment:
No need for the spaces before the `*manager` -- it's not like we can line up with `MMModemFirmwareUpdateMethod` anyway...

Reply:
Done